# Objective


The goal of this project is to independently value a loan portfolio as of **31 December 2020**, in support of the client’s year-end audit. The dataset includes **monthly loan origination amounts** and **observed repayments** for vintages between **June 2019 and December 2020**. 

To estimate the portfolio value, we followed a structured approach:

- **Historical Analysis:** Calculated the actual repayment percentages for each vintage across observed months by dividing repayments by the respective origination amounts.
- **Repayment Curve Modeling:** Derived a 20-month expected repayment curve using the historical averages of repayment behavior across all vintages.
- **Cash Flow Forecasting:** Applied the expected repayment percentages to each vintage’s origination amount to project future monthly cash flows from January 2021 onward.
- **Discounted Cash Flow Valuation:** Used a **5% annual discount rate** (converted to monthly) to discount the forecasted repayments back to their present value as of **31 December 2020**.
- **Variance Analysis:** Compared our model's valuation to the client-provided estimate of **CHF 84,993,122.67**, calculating both absolute and relative differences.
- **Audit Assessment:** Evaluated whether the difference exceeded the audit materiality threshold of **CHF 500,000**.


## Assumptions: 
Customers keep repaying at the same rate as before.

There’s no increase in defaults or economic shocks.

No write-offs or losses are anticipated.

A 5% annual discount rate is used, converted to a monthly rate for DCF purposes.

The discount rate remains constant across the entire forecast period and does not reflect changing market or credit risk conditions.

## Data Cleaning

In [95]:
import pandas as pd
import numpy as np

In [96]:
df=pd.read_csv('Dataset.csv')

In [97]:
df

,Unnamed: 0,Origination Amount,31.05.2019,30.06.2019,31.07.2019,31.08.2019,30.09.2019,31.10.2019,30.11.2019,31.12.2019,...,31.03.2020,30.04.2020,31.05.2020,30.06.2020,31.07.2020,31.08.2020,30.09.2020,31.10.2020,30.11.2020,31.12.2020
0,31.05.2019,10018746.17,1443069.08,3332200.33,1328138.75,928085.74,736418.27,539403.31,427557.86,324459.32,...,116684.68,92699.67,63399.66,53265.12,37121.13,29787.10,24524.90,18085.94,16581.01,11442.97
1,30.06.2019,10868379.04,0.00,1392751.60,3011884.91,1237868.70,970929.28,892351.83,668767.02,505612.59,...,255222.42,198833.96,161996.73,138461.91,92346.68,79641.30,63457.44,52373.85,43374.70,37404.87
2,31.07.2019,10733932.61,0.00,0.00,1537650.24,2953335.55,1208316.08,879375.19,711016.84,658251.40,...,302575.54,258652.52,191798.05,170027.54,127574.33,110301.21,89766.69,64746.84,61408.92,50312.70
3,31.08.2019,12558727.02,0.00,0.00,0.00,1617681.94,4082016.00,1387474.94,1247623.59,886293.35,...,417223.56,336686.08,253556.20,200066.59,151859.74,109973.00,90228.14,70661.50,53102.83,47069.84
4,30.09.2019,14505071.44,0.00,0.00,0.00,0.00,1992242.84,3930445.60,1394620.78,1227905.58,...,628429.48,589692.85,457299.31,323764.87,288152.28,239872.99,192246.98,171550.69,142575.97,116853.05
5,31.10.2019,15652952.20,0.00,0.00,0.00,0.00,0.00,2289453.76,4682354.31,1659503.89,...,763523.36,742787.97,558085.95,461806.22,358671.23,281881.11,241719.91,182730.05,144953.58,119260.10
6,30.11.2019,15107713.30,0.00,0.00,0.00,0.00,0.00,0.00,2162283.09,4637701.69,...,930720.35,697500.94,667277.73,547749.09,387987.02,309448.86,283876.04,215635.85,185516.45,141560.57
7,31.12.2019,17004745.04,0.00,0.00,0.00,0.00,0.00,0.00,0.00,2402403.37,...,1245452.39,1116505.94,803590.21,724956.28,545397.33,458832.95,393971.01,333818.53,286831.44,216447.57
8,31.01.2020,16794379.95,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,...,1505493.21,1182983.95,955821.48,846061.73,683116.81,560572.44,468203.76,349067.92,309854.99,267813.78
9,29.02.2020,19217205.82,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,...,6142911.08,1833677.81,1317065.75,1108494.37,918465.33,719913.69,587381.96,498801.31,371563.77,294941.22


In [98]:
payment_cols = df.columns[2:]   # all calendar payment columns
# mil stands for month-in-life: Refers to how many months have passed since a loan was originated 
mil_rows = []
for _, row in df.iterrows():
    orig = row['Origination Amount']
    payments  = row[payment_cols].values.astype(float)
    first_nz  = next((i for i, v in enumerate(payments) if v > 0), None)
    if first_nz is None:
        mil_rows.append([])
        continue
    mil_pcts = (payments[first_nz:] / orig * 100).tolist()
    mil_rows.append(mil_pcts)
max_mil = max(len(r) for r in mil_rows)
mil_df  = pd.DataFrame(
    [r + [np.nan] * (max_mil - len(r)) for r in mil_rows],
    columns=[f'MIL_{i+1}' for i in range(max_mil)]
)
mil_df['Origination Amount'] = df['Origination Amount'].values
print(mil_df.round(2).to_string())

    MIL_1  MIL_2  MIL_3  MIL_4  MIL_5  MIL_6  MIL_7  MIL_8  MIL_9  MIL_10  MIL_11  MIL_12  MIL_13  MIL_14  MIL_15  MIL_16  MIL_17  MIL_18  MIL_19  MIL_20  Origination Amount
0   14.40  33.26  13.26   9.26   7.35   5.38   4.27   3.24   2.37    1.68    1.16    0.93    0.63    0.53    0.37    0.30    0.24    0.18    0.17    0.11         10018746.17
1   12.81  27.71  11.39   8.93   8.21   6.15   4.65   3.86   3.03    2.35    1.83    1.49    1.27    0.85    0.73    0.58    0.48    0.40    0.34     NaN         10868379.04
2   14.33  27.51  11.26   8.19   6.62   6.13   4.69   3.94   2.82    2.41    1.79    1.58    1.19    1.03    0.84    0.60    0.57    0.47     NaN     NaN         10733932.61
3   12.88  32.50  11.05   9.93   7.06   5.53   4.55   3.32   2.68    2.02    1.59    1.21    0.88    0.72    0.56    0.42    0.37     NaN     NaN     NaN         12558727.02
4   13.73  27.10   9.61   8.47   6.48   5.54   4.33   4.07   3.15    2.23    1.99    1.65    1.33    1.18    0.98    0.81     NaN 

In [ ]:
mil_curve = {}
for col in mil_df.columns[:-1]:     # exclude 'Origination Amount'
    pct    = mil_df[col]
    orig   = mil_df['Origination Amount']
    active = pct.notna()
    mil_curve[col] = (
        (pct[active] * orig[active]).sum() / orig[active].sum()
    ) if active.sum() > 0 else 0.0

expected_frac = [v / 100.0 for v in mil_curve.values()]

print(f"Repayment curve sum: {sum(expected_frac)*100:.2f}%  should be ~95–101%")
for k, v in mil_curve.items():
    print(f"  {k}: {v:.4f}%")


Repayment curve sum: 89.79%  should be ~95–101%
  MIL_1: 14.3069%
  MIL_2: 30.5041%
  MIL_3: 8.0514%
  MIL_4: 6.7191%
  MIL_5: 5.6402%
  MIL_6: 4.6747%
  MIL_7: 3.9617%
  MIL_8: 3.3052%
  MIL_9: 2.7523%
  MIL_10: 2.1659%
  MIL_11: 1.7548%
  MIL_12: 1.4836%
  MIL_13: 1.1426%
  MIL_14: 0.8991%
  MIL_15: 0.7251%
  MIL_16: 0.5588%
  MIL_17: 0.4196%
  MIL_18: 0.3535%
  MIL_19: 0.2585%
  MIL_20: 0.1142%


In [100]:
origination_df    = df.set_index(df.columns[0])
valuation_date    = pd.to_datetime("2020-12-31")
latest_vintage    = pd.to_datetime(origination_df.index[-1], dayfirst=True)
end_date          = latest_vintage + pd.DateOffset(months=len(expected_frac) + 1)
all_payment_dates = pd.date_range(start="2019-06-30", end=end_date, freq="ME")
forecast_matrix   = pd.DataFrame(0.0, index=all_payment_dates, columns=origination_df.index)

for vintage in origination_df.index:
    orig_amount     = origination_df.loc[vintage, "Origination Amount"]
    vintage_date    = pd.to_datetime(vintage, dayfirst=True)
    months_observed = (valuation_date.year  - vintage_date.year)  * 12 + \
                      (valuation_date.month - vintage_date.month)
    for month_idx, frac in enumerate(expected_frac, start=1):
        if month_idx <= months_observed:
            continue    # skip already-observed history
        pay_date = (vintage_date + pd.DateOffset(months=month_idx)).to_period("M").to_timestamp("M")
        if pay_date in forecast_matrix.index:
            forecast_matrix.at[pay_date, vintage] = orig_amount * frac

forecasted_cash_flows_df = pd.DataFrame({
    "Forecasted_Cash_Flow": forecast_matrix.sum(axis=1)
})
forecasted_cash_flows_df = forecasted_cash_flows_df[forecasted_cash_flows_df.index > valuation_date]

preview = forecasted_cash_flows_df.copy()
preview["Forecasted_Cash_Flow"] = preview["Forecasted_Cash_Flow"].apply(
    lambda x: f"CHF {x:,.2f}")
print("Forecasted Cash Flows (Jan 2021 onward):")
print(preview.head(10))


Forecasted Cash Flows (Jan 2021 onward):
           Forecasted_Cash_Flow
2021-01-31    CHF 24,135,875.37
2021-02-28    CHF 20,511,534.38
2021-03-31    CHF 11,718,180.61
2021-04-30     CHF 9,676,083.32
2021-05-31     CHF 7,962,689.44
2021-06-30     CHF 6,518,536.51
2021-07-31     CHF 5,317,465.15
2021-08-31     CHF 4,289,515.99
2021-09-30     CHF 3,423,961.54
2021-10-31     CHF 2,697,227.89


In [101]:
annual_discount_rate  = 0.05
monthly_discount_rate = (1 + annual_discount_rate) ** (1 / 12) - 1

months_diff      = ((forecasted_cash_flows_df.index.year  - valuation_date.year)  * 12 +
                    (forecasted_cash_flows_df.index.month - valuation_date.month))
discount_factors = (1 + monthly_discount_rate) ** (-months_diff)
present_values   = forecasted_cash_flows_df['Forecasted_Cash_Flow'] * discount_factors
portfolio_value  = round(present_values.sum(), 2)

print(f"Final Portfolio Valuation (Discounted to 31 Dec 2020): CHF {portfolio_value:,.2f}")
print(f"Forecast window: {forecasted_cash_flows_df.index.min().date()} to {forecasted_cash_flows_df.index.max().date()}")


Final Portfolio Valuation (Discounted to 31 Dec 2020): CHF 102,291,840.95
Forecast window: 2021-01-31 to 2022-09-30


Variance Analysis & Audit Assessment

In [ ]:
AUDIT_MATERIALITY_THRESHOLD = 500_000       # CHF — per engagement letter

client_estimate = 84_993_122.67
absolute_diff   = round(abs(portfolio_value - client_estimate), 2)

relative_diff_vs_client = round((absolute_diff / client_estimate) * 100, 2)
relative_diff_vs_model  = round((absolute_diff / portfolio_value) * 100, 2)

print("=" * 55)
print("  PORTFOLIO VALUATION — AUDIT VARIANCE REPORT")
print("=" * 55)
print(f"  Model Estimate:                CHF {portfolio_value:>20,.2f}")
print(f"  Client Estimate:               CHF {client_estimate:>20,.2f}")
print(f"  Absolute Difference:           CHF {absolute_diff:>20,.2f}")
print(f"  Relative Diff (vs client):         {relative_diff_vs_client:>19.2f}%")
print(f"  Relative Diff (vs model):          {relative_diff_vs_model:>19.2f}%")
print("=" * 55)

if absolute_diff < AUDIT_MATERIALITY_THRESHOLD:
    print("Within acceptable audit threshold (CHF 500,000).")
else:
    print(f"Exceeds threshold by CHF {(absolute_diff - AUDIT_MATERIALITY_THRESHOLD):,.2f}")
print("=" * 55)


  PORTFOLIO VALUATION — AUDIT VARIANCE REPORT
  Model Estimate:                CHF       102,291,840.95
  Client Estimate:               CHF        84,993,122.67
  Absolute Difference:           CHF        17,298,718.28
  Relative Diff (vs client):                       20.35%
  Relative Diff (vs model):                        16.91%
Exceeds threshold by CHF 16,798,718.28


# Conclusion

Based on a month-in-life aligned repayment curve and a 5% annual discount rate, 
our model independently estimates the portfolio value as of 31 December 2020 to be 
**CHF 102,291,840.95**.

This is higher than the client's estimate of **CHF 84,993,122.67**, resulting in 
an absolute difference of **CHF 17,298,718.28** (+20.35% relative to the client's 
estimate). Since this difference exceeds the **CHF 500,000** audit materiality 
threshold, it indicates a notable discrepancy in the valuation that warrants 
further investigation.

Forecasted repayments are based strictly on observed historical averages across 
all loan vintages, weighted by origination amount and aligned by month-in-life 
position.

The model operates under the following assumptions:
- Customers continue to repay at historical average rates
- No increase in defaults or economic shocks
- No write-offs or losses are anticipated
- A constant 5% annual discount rate applies throughout the forecast period

The remaining gap of **CHF 17.30 million** suggests the client's estimate may 
incorporate one or more of the following adjustments not reflected in our model:

- Expected credit losses or anticipated loan defaults
- A more conservative or alternative repayment curve
- A different discount rate or discounting horizon
- Provisions for charge-offs, delinquency, or economic risk

The model assumes full repayment in line with historical averages, whereas the 
client may have applied credit loss expectations or other risk adjustments.

It is recommended to reconcile assumptions with the client — particularly around 
repayment expectations, loss provisioning, and discounting methodology — to ensure 
alignment for audit validation and financial reporting purposes.
